protected/controllers/CsvController.php

```diff
@@ -210,6 +209,210 @@ private function createNewPatient($patient)
+            }
+            $pat_ref = new PatientUserReferral();
+            $pat_ref->user = $referred_to;
+            if(!$pat_ref->save()){
+                $errors[] = 'Count not save referred to user';
+                array_unshift($errors, $pat_ref->getErrors());
+                return $errors;
+            }
+        }
+
+        //optom
+        if(isset($patient['optom_first_name']) and isset($patient['optom_last_name'])){
+            $optom_label = ContactLabel::model()->findByAttributes(array('name' => 'Optometrist'));
+            $optom_contact = Contact::model()->findByAttributes(array(
+                'contact_label_id' => $optom_label->id,
+                'first_name' => $patient['optom_first_name'],
+                'last_name' => $patient['optom_last_name'],
+            ));
+            if($optom_contact === null){
+                $optom_contact = new Contact();
+                $optom_contact->first_name = $patient['optom_first_name'];
+                $optom_contact->last_name = $patient['optom_last_name'];
+                $optom_contact->label = $optom_label;
+                if(!$optom_contact->save()){
+                    $errors[] = 'Could not save new optometrist contact';
+                    array_unshift($errors, $optom_contact->getErrors());
+                    return $errors;
+                }
+            }
+            $pat_con = new PatientContactAssignment();
+            $pat_con->contact = $optom_contact;
+            if (!$pat_con->save()){
+                $errors[] = 'Could not save optometrist contact assignment';
+                array_unshift($errors, $pat_con->getErrors());
+                return $errors;
+            }
+            $new_gp = new Gp();
+            $new_gp->obj_prof = 0;
+            $new_gp->nat_id = 0;
+            $new_gp->contact = $pat_con;
+            if(!$new_gp->save()){
+                $errors[] = 'Could not save new practitioner contact';
+                array_unshift($errors, $new_gp->getErrors());
+                return $errors;
+            }
+        }
+
+        //opthal
+        if(isset($patient['opthal_first_name']) and isset($patient['opthal_last_name'])){
+            $opthal_label = ContactLabel::model()->findByAttributes(array('name' => 'Ophthalmologist'));
+            $opthal_contact = Contact::model()->findByAttributes(array(
+                'contact_label_id' => $opthal_label->id,
+                'first_name' => $patient['opthal_first_name'],
+                'last_name' => $patient['opthal_last_name'],
+            ));
+            if($opthal_contact === null){
+                $opthal_contact = new Contact();
+                $opthal_contact->first_name = $patient['opthal_first_name'];
+                $opthal_contact->last_name = $patient['opthal_last_name'];
+                $opthal_contact->label = $opthal_label;
+                if(!$opthal_contact->save()){
+                    $errors[] = 'Could not save new ophthalmologist contact';
```

Comment:
You might want to specify the patient that the error was for, and perhaps even store the error message itself. A helper function may aid in this endeavour